### PR TITLE
LandIce: add option to account for the fact that the bed is not flat

### PR DESCRIPTION
**Pull request by mperego**
Add corrective term to velocity and enthalpy basal residual to account for the fact that the tangential velocity to the bed is not the same as the horizontal velocity. Previous behavior can be obtained by setting to true the Problem parameter `Flat Bed Approximation = true`.

Modified tests accordingly.

### CODE REVIEW

File: src/landIce/evaluators/LandIce_StokesFOBasalResid_Def.hpp [hunk 1]
```diff
@@ -40,11 +40,15 @@ StokesFOBasalResid<EvalT, Traits, BetaScalarT>::StokesFOBasalResid (const Teucho
   w_measure = decltype(w_measure)(p.get<std::string> ("Weighted Measure Name"), dl_basal->qp_scalar);
   normals   = decltype(normals)(p.get<std::string> ("Side Normal Name"), dl_basal->qp_vector_spacedim);
```
Comment (bartgol) [inline src/landIce/evaluators/LandIce_StokesFOBasalResid_Def.hpp:41]: We may as well move this inside the if statement too.

File: src/landIce/evaluators/LandIce_BasalMeltRate_Def.hpp [hunk 2]
```diff
@@ -138,8 +147,17 @@ operator() (const Basal_Melt_Rate_Tag& tag, const int& sideSet_idx) const {
...
     //mstar, [W m^{-2}] = [Pa m s^{-1}]: basal latent heat in temperate ice
     ScalarT mstar = geoFluxHeat(sideSet_idx,node);
+    ScalarT tanVelSquared(0.0);
     for (unsigned int dim = 0; dim < vecDimFO; dim++)
-      mstar += beta_scaling * beta(sideSet_idx,node) * velocity(sideSet_idx,node,dim) * velocity(sideSet_idx,node,dim);
+      tanVelSquared += velocity(sideSet_idx,node,dim) * velocity(sideSet_idx,node,dim);
+    if(!flat_approx) {
+      ScalarT tmp(0.0);
+      for (unsigned int dim = 0; dim < vecDimFO; dim++)
+        tmp += -normals(sideSet_idx,node,dim)/normals(sideSet_idx,node,2)*velocity(sideSet_idx,node,dim);
```
[the remark this answers is not in the document]
Comment (bartgol) [reply src/landIce/evaluators/LandIce_BasalMeltRate_Def.hpp:156]: Also, I don't understand where `normals(idim)/normals(2)` factor comes from. I tried to write down some triangles, and see if I could spot the transformatino, but I don't get it.

File: src/landIce/evaluators/LandIce_BasalMeltRate_Def.hpp [hunk 3]
```diff
@@ -47,13 +50,19 @@ BasalMeltRate(const Teuchos::ParameterList& p, const Teuchos::RCP<Albany::Layout
   EnthalpyHs = decltype(EnthalpyHs)(p.get<std::string> ("Enthalpy Hs Side Variable Name"),scalar_layout);
   enthalpyBasalFlux = decltype(enthalpyBasalFlux)(p.get<std::string> ("Basal Melt Rate Variable Name"),scalar_layout);
   basalVertVelocity = decltype(basalVertVelocity)(p.get<std::string> ("Basal Vertical Velocity Variable Name"),scalar_layout);
+  normals   = decltype(normals)(p.get<std::string> ("Side Normal Name"), dl_basal->qp_vector_spacedim);
```
[the remark this answers is not in the document]
Comment (mperego) [reply src/landIce/evaluators/LandIce_BasalMeltRate_Def.hpp]: ouch, fixed

File: src/landIce/evaluators/LandIce_StokesFOBasalResid_Def.hpp (hunk 1 above)
Comment (mperego) [reply src/landIce/evaluators/LandIce_StokesFOBasalResid_Def.hpp]: done

File: src/landIce/evaluators/LandIce_BasalMeltRate_Def.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mperego) [reply src/landIce/evaluators/LandIce_BasalMeltRate_Def.hpp:156]: We solve for the horizontal velocities (parallel to x-y plane), but in the sliding laws and in the frictional heat we should use the velocity tangential to the bed. the expression of the tangential velocity as a  function of the horizontal velocities are obtained using the equation u \dot n = 0 at the bed. See https://www.overleaf.com/project/60a53634001c9b92bba45523, page 6.

**Comment by mperego** [comment]
@bartgol, can I merge this?

**Review by bartgol — APPROVED** [review]

**Pull request merged**